### ISSUE TITLE
Way to publish other files/binaries as GH Release from each build?

### ISSUE THREAD

**Issue by breznak** [joerick/cibuildwheel#363]
I have a c++/py library that uses `python setup.py` to build. 

Would it be possible to extract other files from each of the `cibuildwheel`'s builds (apart from .whl) to the release (`./wheelhouse/`) folder, so I can use them to create a release? 

For example in our current workflow: 
https://github.com/htm-community/htm.core/actions/runs/121495373/workflow 

1. I build once https://github.com/htm-community/htm.core/actions/runs/121495373/workflow#L65 
2. publish the files as a GH Release https://github.com/htm-community/htm.core/actions/runs/121495373/workflow#L72 
3. then I build everything again with cibw: https://github.com/htm-community/htm.core/actions/runs/121495373/workflow#L96 

Desired outcome: 
specify path/pattern of files that are included in "release" after each cibw build. 
Ie:
```
 - name: Build wheel
      env:
        CIBW_INCLUDE_RELEASE: {project}/build/Release/*.[zip,gz]
      run: |
        python -m cibuildwheel --output-dir wheelhouse
```

**Comment by breznak** [comment]
Out of interest, the cibw build is extremely fast (3min, compared to 15min normal CI build time). 
Does it mean it's not entirely re-built again in step 3. (above), but uses the existing binary from 1. if python version matches? 

I'm confused and not sure it builds correctly on the manylinux (?) 
https://github.com/htm-community/htm.core/runs/727475761?check_suite_focus=true#step:13:81

**Comment by YannickJadoul** [comment]
Just to make sure I understand correctly: what exactly do you want to extract extra to the wheel? If you're uploading things to PyPI, a wheel is all people will download, so it would preferably contain everything a release needs to have.
On the other hand, if it's not part of the wheel, building with `cibuildwheel` is probably not the way to go, since it will build everything again for each version of Python.
So, I probably missed some details, so correct me if I'm wrong, but: if it's Python-version-and-platform specific, you probably want to include it in your wheel, and if not, it's probably better to build outside of `cibuildwheel`? Unless it's some kind of shared library independent of the Python version but shared between the different wheels?

> I'm confused and not sure it builds correctly on the manylinux (?)
> https://github.com/htm-community/htm.core/runs/727475761?check_suite_focus=true#step:13:81

If I understand correctly, it doesn't build correctly. I think `auditwheel` is failing, as the old build artifacts (build outside the manylinux image) are being recycled?

Cfr:

```

  Files detected:
    ./build/scripts/src/libhtm_core.so
    ./build/scripts/bindings/py/cpp_src/math.cpython-37m-x86_64-linux-gnu.so
    ./build/scripts/bindings/py/cpp_src/algorithms.cpython-37m-x86_64-linux-gnu.so
    ./build/scripts/bindings/py/cpp_src/engine_internal.cpython-37m-x86_64-linux-gnu.so
    ./build/scripts/bindings/py/cpp_src/sdr.cpython-37m-x86_64-linux-gnu.so
    ./build/scripts/bindings/py/cpp_src/encoders.cpython-37m-x86_64-linux-gnu.so
    ./build/scripts/_CPack_Packages/Linux/TGZ/htm_core-v2.1.6-linux64/lib/libhtm_core.so
    ./build/Release/lib/libhtm_core.so
    ./build/Release/distr/src/htm/dummy.cpython-37m-x86_64-linux-gnu.so
    ./build/Release/distr/src/htm/bindings/math.cpython-37m-x86_64-linux-gnu.so
    ./build/Release/distr/src/htm/bindings/algorithms.cpython-37m-x86_64-linux-gnu.so
    ./build/Release/distr/src/htm/bindings/engine_internal.cpython-37m-x86_64-linux-gnu.so
    ./build/Release/distr/src/htm/bindings/sdr.cpython-37m-x86_64-linux-gnu.so
    ./build/Release/distr/src/htm/bindings/encoders.cpython-37m-x86_64-linux-gnu.so
    ./build/Release/distr/build/lib.linux-x86_64-3.7/htm/dummy.cpython-37m-x86_64-linux-gnu.so
    ./build/Release/distr/build/lib.linux-x86_64-3.7/htm/bindings/math.cpython-37m-x86_64-linux-gnu.so
    ./build/Release/distr/build/lib.linux-x86_64-3.7/htm/bindings/algorithms.cpython-37m-x86_64-linux-gnu.so
    ./build/Release/distr/build/lib.linux-x86_64-3.7/htm/bindings/engine_internal.cpython-37m-x86_64-linux-gnu.so
    ./build/Release/distr/build/lib.linux-x86_64-3.7/htm/bindings/sdr.cpython-37m-x86_64-linux-gnu.so
    ./build/Release/distr/build/lib.linux-x86_64-3.7/htm/bindings/encoders.cpython-37m-x86_64-linux-gnu.so
```

**Comment by breznak** [comment]
> Just to make sure I understand correctly: what exactly do you want to extract extra to the wheel? If you're uploading things to PyPI, a wheel is all people will download 

we're a c++/py project, so on "release" I wish to a) upload a new wheel to pypi, b) create a binary (with the cpp stuff as examples, includes, library files,etc). We do each build with python anyway, but since the python bindings are optional, I need to provide also a "cpp only" install - a typical `tar.gz` archive with the library. 

> If I understand correctly, it doesn't build correctly. I think auditwheel is failing, as the old build artifacts (build outside the manylinux image) are being recycled? 

thank you :+1: yes, that was the case, also why I was wondering if the build is so fast/correct? So it was not, I've now fixed this and do a clean build for CIBW. 

> On the other hand, if it's not part of the wheel, building with cibuildwheel is probably not the way to go, since it will build everything again for each version of Python. 

actually, that would not matter since I have to do all the CIBW builds anyway. If I could access and use the built files from the ciwb builds, it'd still save me time, as now I must do a job for CIBW, and a duplicit non-cibw job that does "the same" but publishes also the files not suitable for a `whl`. 

As seen 
https://github.com/htm-community/htm.core/actions/runs/121585775/workflow#L83 
and non-CIBW
https://github.com/htm-community/htm.core/actions/runs/121585775/workflow#L72 

Sorry for dragging you into specific details, the high-level idea is to access "non-whl files" from/after the CIBW build.

**Comment by YannickJadoul** [comment]
> we're a c++/py project, so on "release" I wish to a) upload a new wheel to pypi, b) create a binary (with the cpp stuff as examples, includes, library files,etc). We do each build with python anyway, but since the python bindings are optional, I need to provide also a "cpp only" install - a typical `tar.gz` archive with the library.

Ah right, I see. So that means the problem is only there on Linux, because of the manylinux Docker images?

But do you then really want to distribute the version that's built inside a `manylinux` image? Depending on your users/intentions, it might be better to build with a newer compiler, independently (and I see you have this ["Install gcc-8" job](https://github.com/htm-community/htm.core/actions/runs/121585775/workflow#L33)).

> actually, that would not matter since I have to do all the CIBW builds anyway. If I could access and use the built files from the ciwb builds, it'd still save me time, as now I must do a job for CIBW, and a duplicit non-cibw job that does "the same" but publishes also the files not suitable for a `whl`.

Would this one extra build (over 4 other builds) then make such a difference?
But yes, I understand more or less now. If feels like there should be a way of compiling less.

However, my main concern is that it might be a good idea to keep building wheels conceptually separate from the binary C++ builds. There might be separate concerns; for example, to build wheels (and Python C extensions) you need to build position-independent code (to link into a dynamic library), while that's maybe not needed for your binaries, etc.

Another thing is more practical: cibuildwheel was not really designed with this in mind. Different wheels produced by cibuildwheel will have different names, but if you copy other artifacts from cibuildwheel, they will possibly overwrite each other after each different version?
And just high-level-wise, I'm not sure how confusing it would be to have `cibuildwheel` build other things than wheels :-/

A possible hack/workaround would be to already build the shared part in `CIBW_BEFORE_BUILD` (since it does not depend on Python, I suspect?) and copy it to `/output`. Not exactly the same, but to save time and not rebuild the shared part for each Python version, I do something similar in my project: https://github.com/YannickJadoul/Parselmouth/blob/f45697c3bae2de1526b27dddfe7115becc5a7e45/.travis.yml#L145

Apart from that, I'll leave this further up to @joerick to check if/how this could fit with the philosophy behind `cibuildwheel`. As you can notice, I'm not really eager to add this and add non-wheel-building functionality into the project :-/

> Sorry for dragging you into specific details, the high-level idea is to access "non-whl files" from/after the CIBW build.

Nono, thank you! This is the level of detail I wanted/needed to judge further on how to approach this.

**Comment by breznak** [comment]
> Ah right, I see. So that means the problem is only there on Linux, because of the manylinux Docker images? 

not really, it's on all platforms, we support users that might not have python at all, so a pure c++ library install. 

> There might be separate concerns; for example, to build wheels (and Python C extensions) you need to build position-independent code (to link into a dynamic library), while that's maybe not needed for your binaries, etc. 

I agree, this package is for people who build wheels (hence have python). 

> Would this one extra build (over 4 other builds) then make such a difference? 

it's 2x the time - I do all-platforms x all-pythons for CPP package, and then all-platforms x all-pythons for CIBW. I'd like to save the CPP part by extracting files from CIBW builds. 

> A possible hack/workaround would be to already build the shared part in CIBW_BEFORE_BUILD (since it does not depend on Python, I suspect?) and copy it to /output. 

I think this is not what I really want - this saves time from CIBW by re-using some built code, I want the other way around, take code/binaries from CIBW build to reuse/publish it. 

> As you can notice, I'm not really eager to add this and add non-wheel-building functionality into the project :-/

Agree, I wanted to know if you see some relatively easy way to hack-in this behavior. Eg something like 
```
CIBW_TEST_COMMAND="cp -a build/ /some/shared/path" (but I cannot get the shared path in the CI jobs)
```

A way that I imagine would work is using upload-artifact internally: 
```
- uses: actions/upload-artifact@v2
  with:
    name: "cibw"
    path: <xxx>
```
and adding another CIBW option as 
```
CIBW_ARTIFACT_PATTERN="/build/Release/*[so,dll,exe]"
```
This patter would be replaced for xxx above and uploaded as an Artifact. 

Issues: 
- files could be overriden (but users can counter that in setup.py and create unique names where needed) 
- this would work for GH Actions, I don't know about other supported CI
- I guess it would be good for some use-cases, but your job is to keep the project in line.

**Comment by henryiii** [comment]
I believe this is how you upload:

https://scikit-hep.org/developer/gha_wheels#publishing

(Will be tested on the next release of boost-histogram)

You just need a upload artifacts (useful anyway for checking the results if you haven't published yet), it makes sure you have no errors in any of the builds (no partial uploads), and is a good place to make the sdist. I don't think there's much to be gained in making the upload action multi platform or adding the ability to skip existing; this is a good workflow.

Also see setuptools_scm for a nice example of an upload. Be sure to use v2 of upload and download actions.

**Comment by joerick** [comment]
We've seen a similar feature request before - #115 discusses how we might design a feature to extract coverage reports from builds.

The challenge is that the build runs so many times, so how not to copy extracted files over one another.

It could work something like:

```yaml
CIBW_OUTPUT_DIR=dist
CIBW_EXTRA_ARTIFACTS="build/Release/*.so coverage.json"
```

Resulting in the following directory structure once the build is complete:

```txt
dist
 ├─ cp27-manylinux2010_x86_64.whl
 ├─ cp35-manylinux2010_x86_64.whl
 ├─ cp36-manylinux2010_x86_64.whl
 ├─ cp37-manylinux2010_x86_64.whl
 └─ extra_artifacts
    ├─ cp27-manylinux_x86_64
    │  ├─ build
    │  │  └─ Release
    │  │     ├─ mylib.so
    │  │     └─ deplib.so
    │  └─ coverage.json
    ├─ cp35-manylinux_x86_64
    │  ├─ build
    │  │  └─ Release
    │  │     ├─ mylib.so
    │  │     └─ deplib.so
    │  └─ coverage.json
    ├─ cp36-manylinux_x86_64
    │  ├─ build
    │  │  └─ Release
    │  │     ├─ mylib.so
    │  │     └─ deplib.so
    │  └─ coverage.json
    └─ cp37-manylinux_x86_64
       ├─ build
       │  └─ Release
       │     ├─ mylib.so
       │     └─ deplib.so
       └─ coverage.json
```

Feels kinda messy, but it could work. Would that help for your use case @breznak ?

**Comment by breznak** [comment]
> Feels kinda messy, but it could work. Would that help for your use case 

it's a while, sorry..but looking at it, yes, that would help! :+1: 

> Resulting in something like the following directory structure once the build is complete 

It could be like you suggest, I think my idea was to simply add a `<build-job>.tar.gz`
file. 
SO 
```
dist
 ├─ cp27-manylinux2010_x86_64.whl
 |-   cp27-manylinux2010_x86_64.zip
 ├─ cp35-manylinux2010_x86_64.whl
 |-  cp35-manylinux2010_x86_64.zip 
```

> The challenge is that the build runs so many times, so how not to copy extracted files over one another. 

yes, I don't see a problem with overwriting if each job-type (cp27-manylinux2010_x86_64, cp35-manylinux2010_x86_64,...) is used as a prefix for the resulting archive/subfolder.

**Comment by YannickJadoul** [comment]
Sorry, this thing completely slipped my mind. Also partially because I couldn't really find a way to do this.

I'm still a bit hesitant about such a change, because it feels like it could be reasonably much code for a feature that's not really the goal of `cibuildwheel` (i.e., building wheels).

But if @joerick deems this is fine to add, I suppose we could try.

Another idea: would it make sense to copy out changes to the project back out of the docker image? That way, `cibuildwheel` is more transparent compared to Windows/macOS, where the resulting build is still available after `cibuildwheel`.

> it's 2x the time - I do all-platforms x all-pythons for CPP package, and then all-platforms x all-pythons for CIBW. I'd like to save the CPP part by extracting files from CIBW builds.

I'm still not really getting this, btw. Why do you need "all-platforms x all-pythons for CPP package"; isn't the C++ code Python-independent, then?

**Comment by larsoner** [comment]
FWIW in my `CIBW_BEFORE_ALL` I am doing (when on Linux):
```
mkdir -p /output
cp my_cool_file /output/
```
inspired by [the python code doing this sort of thing](https://github.com/pypa/cibuildwheel/blob/8537b1670f47e358d26afb035c0818329f78429d/cibuildwheel/linux.py#L120), and it seems to work to get things to show up in the `output-dir` of the host. You could probably also do this in your TEST step for example, but I haven't tried.

At the moment this is probably abusing the system a bit and isn't guaranteed to work in the future in any way -- but I figured I'd at least mention it here in case anyone else wants to take their chances 🙂

(FWIW, my use case is that I want to build separate binary-only installers for a C++ package for which I *also* -- in a separate GH action -- build a PyPi wheel without the binaries, and `cibuildwheel` provides a really nice interface for this.)

> I'm still a bit hesitant about such a change, because it feels like it could be reasonably much code for a feature that's not really the goal of cibuildwheel (i.e., building wheels).

Agreed, allowing people to specify what to copy and when seems like a pain. Perhaps an alternative would to expose a new `CIBW_CONTAINER_OUTPUT_DIR` env var (which I guess would be `'/output'` for now) -- for the "output directory within the build container that gets mapped to the output-dir of the host (only on Linux)". It seems like it would satisfy many general use cases, and not require much code at the `cibuildwheel` end. (At least I assume exposing a new env var is easy enough!)
